<a href="https://colab.research.google.com/github/harrisb002/CS_415/blob/Midterm1/Midterm1_Solutions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# My personal solutions to Midterm 1


# Problem 1:
- Consider the algorithm divide shown below:


In [7]:
def divide(x, y):
  """ Input:  Two n-bit integers x and y, where y>=1
      Output: The quotient and remainder of x divided by y
  """
  if x == 0:
      return (0,0)

  (q,r) = divide(x//2,y)
  q=2*q
  r=2*r
  if x % 2 != 0:
      r = r+1
  if r >= y:
      r = r-y
      q = q+1
  return (q, r)

(a) Show the successive recursive calls made and the results returned by each call and the final output for the input pair a = 121 and b = 26
- Initial Call: divide(121, 26) -> x = 121 (odd)
- Rec Call 1: divide(60, 26) -> x = 60 (even)
- Rec Call 2: divide(30, 26) -> x = 30 (even)
- Rec Call 3: divide(15, 26) -> x = 15 (odd)
- Rec Call 4: divide(7, 26) -> x = 7 (odd)
- Rec Call 5: divide(3, 26) -> x = 3 (odd)
- Rec Call 6: divide(1, 26) -> x = 1 (odd)
- Base Case: divide(0, 26) -> returns q = 0, r = 0

Unraveling:
- 7) divide(1, 26) -> q = 0, r = 1 (3 is odd, r+1)
- 6) divide(3, 26) -> q = 0, r = 3 (3 is odd, r+1)
- 5) divide(7, 26) -> q = 0, r = 7 (7 is odd, r+1)
- 4) divide(15, 26) -> q = 0, r = 15 (15 is odd, r+1)
- 3) divide(30, 26) -> q = 0, r = 30 -> q = 1, r = 4 (30 is even)
- 2) divide(60, 26) -> q = 2, r = 8 (60 is even)
- 1) divide(121, 26) -> q = 4, r = 17 (121 is odd, r+1)

Final result (4, 17)


(b) How many recursive calls are made by divide for the input a = 2^100   and b = 347, not counting the original call?
- ANS: The number of times 2^100 needs to be halved until it reaches 0 is equivalent to its binary representation length. In binary,  2^100   is represented as a 1 followed by 100 zeros.  Therefore, it needs to be halved 100 times to reach 1, and then one more time to reach 0. Using this as evidence and not counting the original call,  the number of recursive calls (excluding the original) for the input a = 2^100  and b = 347 is  = 100 times.

(c) For the input b = 2^100 - 1 , and a in the range  2^100 ⩽  a ⩽ 2^101 how many values of ‘a’ will result in skipping both “if” statements? How many will cause execution of the code associated with both “if” statements? (In the former case, x is even and r < y while in the latter x is odd and r ⩾ y
- ANS: “Skipping both “if'' statements,  when x is even and r < y:  Given the range of ‘a’, half of the numbers will be even,. We can definitively say that a = 2^100 will skip both “if statements”, however, just being even is not enough for other numbers in the range. As the condition also requires r < y which means the remainder, after being doubled (and incremented if odd) must be less than y. But, due to the halving nature of the recursive calls and how r is computed, it's obvious that the remainder r will surpass y after several recursive calls. Given the size of y (which is 2^100 - 1) it's extremely unlikely for r to be less than y after a few recursive calls as the value of r increases exponentially with each call. Thus in this range only a = 2^100   will skip both “if” statements.

- ANS: Executing both “if” statements, when x is odd and r ⩾ y: For x to be odd, a itself should be odd. In the given range every number that isn't a power of 2 will be odd. This excludes 2^100 and 2^101 but includes all numbers in between. However, given the size of y (2^100 - 1) after a few recursive calls, r is bound to become greater than or equal to y do to the nature of the algorithm. This is because, with each recursive call, r is at least doubled. Even if the initial r values are small, they will grow rapidly and will eventually surpass y. Thus, in this range, all odd values of a except 2^100 and 2^101 will execute the code associated with both “if” statements. Therefore, No values of a will skip both “if” statements and all odd values of a in the given range, excluding 2^100 and 2^101  will execute the code associated with both “if” statements.

(d) Modify the algorithm so that no recursive calls are made when x ⩽ y or y ⩽ 2. Rewrite the complete code with suitable changes.
- If x is less than or equal to y, it means the quotient is 0, and the remainder is x.
- If y is less than or equal to 2, compute the quotient and remainder using simple division and modulo operator without recursion.
- If these base cases are not met then the function continues recursion division as per usual.


In [ ]:
  def divide(x, y):
    """ Input: Two n-bit integers x and y, where y>=1
        Output: The quotient and remainder of x divided by y
    """
    # Handle base cases
    if x <= y:
        return (0, x)
    if y <= 2:
        return (x // y, x % y)
    if x == 0:
        return (0, 0)

    # Recursive division
    (q, r) = divide(x // 2, y)
    q = 2 * q
    r = 2 * r
    if x % 2 != 0:
        r = r + 1
    if r >= y:
        r = r - y
        q = q + 1
    return (q, r)

# Problem 3:
- Shown below is the extended-Euclid’s algorithm to find the gcd(d) of a given pair of integers ‘a’ and ‘b’ and to find integers x and y such that  ax + by = d (Note that this algorithm assumes a >= b >= 0)


In [ ]:
def gcdext(a,b):
  """ Input:  Two integers a and b with a>=b>=0
      Output: Integers x,y,d such that d=gcd(a,b) and ax+by=d
  """
  if b==0:
      v = (1, 0, a)
      return v
  (x1,y1,d) = gcdext(b,a%b)
  v = (y1,x1-a//b*y1, d)
  return v

(a) For a = 139, find a value of b for which the algorithm will terminate after three recursive calls. (For example, if b = 69, it terminates after two recursive calls: (139, 69).
With a = 139, which choice of b (in the range will result in a maximum number of recursive calls? You can solve this problem by writing a program or by a deductive approach.
- First Part: If we choose a value that if doubled is 1 more than 139, such as 70, then we know that the remainder will be 1 after the second recursive call and therefore need one more to terminate at 0. Thus by choosing b = 70 we get:
139 % 70 =  70, 69
70 % 69 = 69, 1
69 % 1 = 0

ANS: Thus b = 70 meets this condition of 3 recursive calls

- Second Part: Using the programmatic method. The purpose of this function is to iterate over all possible values of b (from 1 to a−1) and find out which ‘b’ causes the most recursive calls when finding the greatest common divisor (gcd) of a and b using the extended Euclid’s algorithm.

ANS: Using this algorithm below I found that b = 86 with 9 recursive calls is the max.

In [ ]:
def test_gcdext(a):
    # Counts
    max_calls = 0
    b_max = 0

    for b in range(1, a):
       # Using count as a list with a single element
       # so it is mutable and can be modified within gcdext_modified
       # and its changes are seen outside the function as well
        count = [0]

        # Basically same as gcd_ext but with a count
        def gcdext_modified(a, b):
            # Incrementing the value inside the list to see which has most rec calls
            count[0] += 1
            if b == 0:
                return (1, 0, a)
            (x1, y1, d) = gcdext_modified(b, a % b)
            return (y1, x1 - a // b * y1, d)

        gcdext_modified(a, b)

        # After the function completes for a particular b, if the count (# of rec.calls) is more than
        # the previous maximum (max_calls), the max_calls and b_max values are updated.
        if count[0] > max_calls:
            max_calls = count[0]
            b_max = b

    return b_max, max_calls

print(test_gcdext(139))

(86, 9)


(b) Exhibit the successive recursive calls made on the input pair where (2^m + 2^n, 2^n) where (0 < m < n)  and all the outputs returned by all the calls, and the final output. Repeat the same problem for the input pair  (2^n + 1,  2^n-1 + 1) where (0 < m < n)
- a = 2^m + 2^n  , b = 2^n
Since 2^m is a power of 2, any number less than 2^n but greater than 2n-1 will have a remainder of the number itself when divided by 2^n. Therefore, (2^m + 2^n) mod 2^n = 2^m because simply adding 2n with a modulo of 2^n will cancel itself leaving 2^m.
Resulting in (2^n, 2^m) and for this pair since n > m, 2^n is a multiple of 2m so the remainder would be 0 as the larger number, i.e. 2^n would be wholly divisible by the smaller number 2m leaving no remainder. Thus ending the recursion with 0.
Therefore the outputs would resemble:
For (2^n, 2^m) -> (0,1,2^m) because when b = 0, x =1, y = 0, and d = a
For (2^m + 2^n, 2^n) =>  (1, 0, 2^m)

Final output is  (1, 0, 2^m)

- a = 2^n + 1, b = 2^n-1 + 1

1. First call: gcdext(2^n + 1, 2^(n-1) + 1)
     a = 2^n + 1, b = 2^(n-1) + 1
        gcdext(2^(n-1) + 1, (2^n + 1) % (2^(n-1) + 1))
        (2^n + 1) % (2^(n-1) + 1)  =>  2^(n-1)

2. Second call: gcdext(2^(n-1) + 1, 2^(n-1))
     a = 2^(n-1) + 1, b = 2^(n-1)
       gcdext(2^(n-1), (2^(n-1) + 1) % 2^(n-1))
       (2^(n-1) + 1) % 2^(n-1)  => 1

3. Third call: gcdext(2^(n-1), 1)
    a = 2^(n-1), b = 1
        gcdext(1, 2^(n-1) % 1)   =>  gcdext(1,0)

4. Fourth call: gcdext (1,0)
     a = 1, b = 0   =>  (1, 0, 1)

	In Returning:

3. Third call return:
    (y1,x1-a//b*y1, d) => (0, 1-2^(n-1)//1*0, 1) => (0,1,1)

2. Second call return:
    * (y1,x1-a//b*y1, d) => (1, 0-(2^(n-1) + 1)//2^(n-1)*1, 1)  =>  (1,-1,1)

1. First call return:
    * (y1,x1-a//b*y1, d)  =>  (-1, 1-(2^n + 1)//(2^(n-1) + 1)*(-1), 1)  =>  (-1,2,1)

Final output is (-1,2,1).


(c) Exhibit a specific input pair (a, b) for which extended-Euclid will make at least 100 recursive calls to compute the output with the additional condition that a and b can’t have more than 25 digits. (Hint: You may find some help with a homework problem.) Prove that your answer is correct. (You can run a program or provide a mathematical proof.)

- A key idea is to use the fact that when using the Euclidean algorithm on two consecutive fibonacci numbers, fib(n) and fib(n-1), the first reduction is always to fib(n-2). Clearly stated: fib(n) % fib(n-1) = fib(n-2)   thus making it a worst case scenario situation regarding the extGcd algorithm because the numbers are decreasing at the slowest rate possible. By choosing  a = fib(100) and b = fib(101) this will in turn need to perform 101 recursive calls (including the initial call). This can be demonstrated by this simple algorithm to count the number of recursive calls using these parameters.

ANS: a = fib(100) and b = fib(101) (which are both less than 25 digits long)


In [ ]:
def fib(n):
  if n == 0:
      return 0
  f = [x for x in range(n+1)]
  f[0] = 0
  f[1] = 1
  for i in range(2,n+1):
      f[i] = f[i-1] + f[i-2]
  return f[n]


rec_calls_count = 0
def extended_gcd_count(a, b):
    global rec_calls_count
    rec_calls_count += 1
    if b == 0:
        return (a, 1, 0)
    else:
        (d, x, y) = extended_gcd_count(b, a % b)
        return (d, y, x - (a // b) * y)

extended_gcd_count(fib(100), fib(101))
print(rec_calls_count)

101


# Problem 4
- Write a function Error-Prob3 that takes as input an integer N and k and outputs the exact error probability associated with the modified primality algorithm implemented in project 1. The idea is as follows: Given as input an integer N and confidence parameter k, first test if N is divisible by 2, 3, 5, 7 or 11. If it is divisible by any of these numbers output(no); else call primality2 (Figure 1.8, page 27) with N and k as inputs. (This in turn calls primality algorithm (Figure 1.7) that randomly chooses a (where 1 < a < N) and tests if a N−1 ≡ 1(mod N) and repeats it k times.) Note that the primality algorithm in Figure 1.7 chooses an a < N which includes the choice of a = 1. For this problem, we will exclude 1 and assume 1 < a < N.

- Your submission shall include the code for Error-Prob3 as well as the output probabilities for the inputs 1801,
2046 and 9571, with k = 4 in all cases.

- Note that 2 is also included in the above divisibility test since the primality test allows generation of an even number for a. (But I excluded it in the project since our random number generator excluded odd numbers.)


In [ ]:
def Error_Prob3(N, k):
    # First check If N is divisible by any of 2, 3, 5, 7, or 11, then error probability is 0.
    for prime in [2, 3, 5, 7, 11]:
        if N % prime == 0:
            return 0

    # Count the number of 'a' for which N appears prime
    count = sum([1 for a in range(2, N) if pow(a, N-1, N) == 1])

    # Calculate error probability for one loop
    error_prob_one_loop= count / (N-1)

    # Calculate error probability for k loops
    error_prob_k = error_prob_one_loop

    return error_prob_k

# Test for given inputs, make k = 4
for n in [1801, 2046, 9571]:
    print(f"Error probability for N={n}: {Error_Prob3(n, 4)}")


Error probability for N=1801: 0.9994444444444445
Error probability for N=2046: 0
Error probability for N=9571: 0.00031347962382445143


- Explanation: This function is used to find the probability that a composite number N would mistakenly appear as prime after performing the test k times. Just as in the project, we first eliminate numbers that are divisible by small primes: 2, 3, 5, 7, and 11.
- If N is divisible by any of these primes, it's not prime. In these cases the function returns an error probability of 0 since the number will be correctly identified as composite such as 2048 because it is divisible by 2 and thus return a 0.
- We can loop  through every possible ‘a’ in the range (2, N) and count how many of these values make N appear prime by satisfying the condition and the function checks this by computing an-1 mod N == 1 for each. To find the error probability for a single loop/test the function gets the fraction of ‘a’ values that make N appear prime out of the possible N-1 values.
- If this fraction is high, it means there are many numbers ‘a’ which would incorrectly verify N as a prime. After using the test k times, the error probability ‘a’ is obtained by raising the single iteration error probability to the power of k. It is staggering how high the probability (almost 99.78%) of falsely identifying 1801 is as a prime for any single iteration of the Fermat primality test.
- However, even though there is this high of an error probability the more times you run the test (i.e., with different random values of a), the more the error will decrease and it should be noted that if given a high enough k value the chance of a false prime identification would be significantly lower.

# Problem 5
- Show the steps involved when Karatsuba’s (divide and conquer based multiplication) algorithm is run with inputs x = 52 and y = 69. Start with 8-bit binary representations of x and y. Exhibit the inputs to the three subproblems generated. Assuming that the correct outputs for these three subproblems are returned by the calls, show the combining step that uses the outputs of the three recursive calls and exhibit the final output. Your final answer should be a 16-bit representation of the product x y. (Show all the computations in binary.)

- X = 00110100  (52)
- Y = 01000101  (69)
- Xh = 0011  (3)    Xl = 0100  (4)   Yh = 0100  (4)   Yl = 0101 (5)

- P1 = Xh  * Yh = 0011 * 0100 = 1100       (12)
- P2 = Xl  * Yl  = 0100 * 0101 = 10100     (20)
- P3 = (Xh  + Xl) * (Yh  * Yl ) = (0011 + 0100) * (0100 + 0101) = 0111 * - 1001 = 111111    (63)

- Result = P1 * 28  + (P3 - P1 - P2) * 24 + P2
- (P3 - P1 - P2) = 111111 − 1100 = 110011 - 10100 = 11111  (63 - 20 - 12 = 31)
- Substitute back in:
- 110000000000 + 111110000 + 10100 = 0000111000000100 (52 * 69 = 3588)

ANS: 0000111000000100


# Problem 6:
- Consider the following modular exponentiation algorithm:


In [ ]:
def modexp(x,y,N):
  """ Input:  Two n-bit integers x and N, an integer exponent y
      Output: x^y mod N
  """
  if y == 0:
      return 1
  z = modexp(x, y//2, N)
  if y % 2 == 0:
      return (z * z) % N
  else:
      return (x * z * z) % N

(a) What is the number of mod operations performed by this algorithm on the input and
x =72, y = 29 amd N = 1321 ?

- For y = 29:
y is odd, so it'll take the path return (x * z * z) % N.     => 1 Mod Operation
It also calls  modexp(x, y//2, N),  =>  modexp(72, 14, 1321).

- For y = 14:
y is even, so it'll take the path return (z * z) % N.      => 1 Mod Operation
It also calls modexp(x, y//2, N),   =>  modexp(72, 7, 1321).

- For y = 7:
y is odd, so it'll take the path return (x * z * z) % N.     => 1 Mod Operation
It also calls modexp(x, y//2, N),   =>   modexp(72, 3, 1321).

- For y = 3:
y is odd, so it'll take the path return (x * z * z) % N.     => 1 Mod Operation
It also calls modexp(x, y//2, N),  =>   modexp(72, 1, 1321).

- For y = 1:
y is odd, so it'll take the path return (x * z * z) % N.     => 1 Mod Operation
It also calls modexp(x, y//2, N),   =>   modexp(72, 0, 1321).

- For y = 0:
The function just returns 1 without any mod operation.

- So, counting the mod operations for each of the above steps:
1 + 1 + 1 + 1 + 1 = 5

ANS: 5 mod operations


(b) What is the number of multiplications performed on the input ?(Note that z2 mod n involves one multiplication and x * z2) involves two multiplications.

For even y: (z * z) = 1
For odd y: (x * z * z) = 2

- For y = 29 (odd)   => 2 Multiplications
modexp(72, 14, 1321).

- For y = 14 (even)  => 1 Multiplication
modexp(72, 7, 1321).

- For y = 7 (odd)  => 2 Multiplications
modexp(72, 3, 1321).

- For y = 3 (odd)  => 2 Multiplications
modexp(72, 1, 1321).

- For y = 1 (odd)    => 2 Multiplications
modexp(72, 0, 1321).

- For y = 0:
The function just returns 1 without any multiplication.

- Adding up the multiplications for each of the above steps:
2 + 1 + 2 + 2 + 2 = 9

ANS: 9 mod operations


(c) Among the n-bit integers, which input invokes the maximum number of mod operations? Which input invokes the least number of MOD operations? What are the number of multiplications in these two cases?

The number of operations depends on how many times the exponent `y` can be halved before reaching 0.

1. Maximum number of mod operations:
This will occur for the largest `y`, which is (2^n - 1), since it's an n-bit integer where all bits are set to 1. This is because the largest number of recursive calls are made when `y` is odd in each recursion (as even `y` would halve more quickly). For a number with all bits set to 1, it will result in the most recursive calls.
   
Number of multiplications:
Since (2^n - 1) is odd, every recursive call results in 2 multiplications. The depth of recursion is `n`, and thus there will be (2n) multiplications.

2. Least number of mod operations:
This will occur for the smallest non-zero `y`, which is 1, since this directly leads to just   one recursive call (with `y=0`) and no further recursive calls.

Number of multiplications:
For `y=1`, there's only one recursive call leading to `y=0`. In this case, there are 2    multiplications for the base case when `y=1`.


(d) (This question is not related to the modexp algorithm) Let N = 101^4 . How many integers a, 1 < a < N have inverses modulo N ?

- An integer ‘a’a has a modular inverse modulo N if and only is gcd(a, N) = 1 (Basically ‘a’ Has an inverse modulo N if and only if ‘a’a is relatively prime to N. To count the number of integers less than N that are coprime to N, you can use the Euler's totient function where pk = pk - pk-1 and therefore (1014) = 1014 - 1013
This results in 103050100
